cpu/sam3x8e/periph/pwm.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (C) 2015 Hamburg University of Applied Sciences
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_sam3x8e
+ * @{
+ *
+ * @file
+ * @brief       CPU specific low-level PWM driver implementation for the SAM3X8E
+ *
+ * @author      Andreas "Paul" Pauli" <andreas.pauli@haw-hamburg.de>
+ *
+ * @}
+ */
+#include <stdint.h>
+#include "board.h"
+#include "periph/pwm.h"
+
+/*
+ * guard file in case no PWM device is defined,
+ */
+#if PWM_NUMOF && PWM_0_EN
+
+#define ERR_INIT_MODE           (-1)
+#define ERR_INIT_BWTH           (-2)
+#define ERR_SET_CHAN            (-1)
+
+/*
+ *
+ */
+int pwm_init(pwm_t dev, pwm_mode_t mode,
+             unsigned int frequency,
+             unsigned int resolution)
+{
+
+    int32_t retval = ERR_INIT_MODE;    /* Worst/First case */
+    uint32_t pwm_clk = 0;              /* Desired/real pwm_clock */
+    uint32_t diva = 1;                 /* Candidate for 8bit divider */
+    uint32_t prea = 0;                 /* Candidate for clock select */
+
+    PwmCh_num *chan;
+
+    /*
+     * Mode check.
+     * Only PW_LEFT -which is hw default- supported for now.
+     */
+    switch (mode) {
+        case PWM_LEFT:
+            break;
+
+        case PWM_RIGHT:
+        case PWM_CENTER:
+        default:
+            return ERR_INIT_MODE;
+    }
+
+    /* Should check if "|log_2 frequency|+|log_2 resolution| <= 32" */
+    pwm_clk = frequency * resolution;
+
+    /*
+     * The pwm provides 11 prescaled clocks with (MCK/2^prea | prea=[0,10])
+     * and a divider (diva) with a denominator range [1,255] in line.
+     * With every step the prescaler is doubled and the remaining quotient
+     * is halfed and tested <=255.
+     */
+    if (F_CPU < pwm_clk) {  /* Have to cut down resulting frequency. */
+        frequency = F_CPU / resolution;
+    }
+    else {   /* Estimate prescaler and divider. */
+        diva = F_CPU / pwm_clk;
+
+        while ((prea < MCK_DIV_LB_MAX) && (~0xff & diva)) {
+            prea = prea + 1;
+            diva = diva >> 1;
+        }
+
+        frequency = F_CPU / ((resolution * diva) << prea);
+    }
+
+    retval = frequency;
+
+    /* Activate PWM block by enabling it's clock. */
+    PMC->PMC_PCER1 = PMC_PCER1_PID36;
+
+    /*
+     * ++ Initialisation ++
+     * adviced by "38.6.5.1 Initialization" in
+     * "Atmel SAM3X8E  SAM3X8C SAM3X4E SAM3X4C SAM3A8C SAM3A8C Datasheet"
+     */
+    /* Unlock User Interface */
+    PWM_0_DEV->PWM_WPCR  = PWM_ENA_CHID0;
```

Comment:
O.K., got it. I focused to much on `pwm_t dev` as a selector, in case there are more than one. So I lost the aspect of whitelisting.
I will (re)introduce the switches in all functions.

Anyhow, your above example wouldn't compile, as long  as nobody places `#define PWM_2_EN 1` in _arduino-due/include/periph_conf.h_. And no one would be so insane to do that. ;)